src/Cello/control_new_refresh.cpp

```diff
@@ -0,0 +1,932 @@
+
+  ItNeighbor it_neighbor =
+    this->it_neighbor(min_face_rank,index_, neighbor_leaf,0,0);
+
+  int il=0;
+
+  int if3[3];
+  while (it_neighbor.next(if3)) {
+
+    const int level_face = it_neighbor.face_level();
+
+    int ic3[3];
+    it_neighbor.child(ic3);
+
+    const int refresh_type = 
+      (level_face == level - 1) ? refresh_coarse :
+      (level_face == level)     ? refresh_same :
+      (level_face == level + 1) ? refresh_fine : refresh_unknown;
+
+    int index_lower[3] = {0,0,0};
+    int index_upper[3] = {1,1,1};
+    refresh->get_particle_bin_limits
+      (rank,refresh_type,if3,ic3,index_lower,index_upper);
+
+    // ASSERT: il < nl
+    particle_determine_periodic_update_
+      (index_lower,index_upper,&dpx[il],&dpy[il],&dpz[il]);
+
+    il++;
+
+  }
+
+  ParticleDescr * p_descr = cello::particle_descr();
+
+  // Apply the updates to the list of particles
+
+  for (int il=0; il<nl; il++) {
+
+    ParticleData * p_data = particle_list[il];
+    Particle particle_neighbor (p_descr,p_data);
+
+    if ( ((rank >= 1) && dpx[il] != 0.0) ||
+	 ((rank >= 2) && dpy[il] != 0.0) ||
+	 ((rank >= 3) && dpz[il] != 0.0) ) {
+	
+      // ... for each particle type
+      const int nt = particle_neighbor.num_types();
+      for (int it=0; it<nt; it++) {
+
+	// ... for each batch of particles
+	const int nb = particle_neighbor.num_batches(it);
+	for (int ib=0; ib<nb; ib++) {
+
+	  particle_neighbor.position_update (it,ib,dpx[il],dpy[il],dpz[il]);
+
+	}
+      }
+    }
+  }
+}
+//----------------------------------------------------------------------
+
+void Block::particle_scatter_neighbors_
+(int npa,
+ ParticleData * particle_array[],
+ std::vector<int> & type_list,
+ Particle particle)
+{
+  const int rank = cello::rank();
+
+  //     ... get Block bounds 
+  double xm,ym,zm;
+  double xp,yp,zp;
+  lower(&xm,&ym,&zm);
+  upper(&xp,&yp,&zp);
+
+  // find block center (x0,y0,z0) and width (xl,yl,zl)
+  const double x0 = 0.5*(xm+xp);
+  const double y0 = 0.5*(ym+yp);
+  const double z0 = 0.5*(zm+zp);
+  const double xl = xp-xm;
+  const double yl = yp-ym;
+  const double zl = zp-zm;
+
+  int count = 0;
+  // ...for each particle type to be moved
+
+  for (auto it_type=type_list.begin(); it_type!=type_list.end(); it_type++) {
+
+    int it = *it_type;
+
+    const int ia_x  = particle.attribute_position(it,0);
+
+    // (...positions may use absolute coordinates (float) or
+    // block-local coordinates (int))
+    const bool is_float = 
+      (cello::type_is_float(particle.attribute_type(it,ia_x)));
+
+    // (...stride may be != 1 if particle attributes are interleaved)
+    const int d  = particle.stride(it,ia_x);
+
+    // ...for each batch of particles
+
+    const int nb = particle.num_batches(it);
+
+    for (int ib=0; ib<nb; ib++) {
+
+      const int np = particle.num_particles(it,ib);
+
+      // ...extract particle position arrays
+
+      double xa[np],ya[np],za[np];
```

Comment:
`np` not known at compile time.
```suggestion
      std::vector<double> xa(np) ,ya(np) ,za(np) ;
```
Likely also requires update to the `position` call below.